### DISCUSSION TITLE
Is it possible to "compress" migrations?

### DISCUSSION THREAD

**Discussion by artsiommiksiuk** [mikro-orm/mikro-orm#5596] [Q&A]
Overtime there are a lot of migrations files can appear, but I don't want to store all individual migration files, and instead do "compression" or squashing cycles to collect all existing migrations and generate only one migration file, which can act as a new initial migration file.

I thought about ensuring ensuring all "production" applied migrations to be applied locally, remove them and then generate via `npx mikro-orm-esm --config src/db/cli.config.ts migration:create --initial` "new" initial migration, but it refuses to do so, because `Error: Initial migration cannot be created, as some migrations already exist`.

Is there by any chance such common practice or tool for that and I just overlooked it?

It doesn't make sense to have version control on top of version control, and applied individual migration files becoming useless if we can squash them into one which allows us to have base database state.

**Comment by B4nan** [comment]
As long as you did not alter the migrations by hand, you can drop the migrations folder and table and start over with the `--initial` flag. It will use the SQL dump produced by `schema:create`, which is completely based on the entity metadata. If you modified some migrations by hand, you could lose those changes, so it's better to do this manually (you could still use the output of `schema:create` and modify it to match what you want).

There is no "compress" command because doing this is non-trivial for all possible scenarios. People can define imperative migrations (e.g. using `this.execute()` instead of `this.addSql()`), and those can't be merged easily, we would need to merge the code of the separate migration files one by one).

**Reply by artsiommiksiuk** [reply]
But that's the point, I can't use --inital, it says there are some migrations already exist. I'm completely dropping the state file and all migrations files, so I'm assuming it should try to read it from db, but it not working like that.

**Reply by artsiommiksiuk** [reply]
Ah, my bad, I need to drop table as well.

**Reply by artsiommiksiuk** [reply]
@B4nan, but actually one question, if I drop table locally and recreate all migrations, wouldn't it then conflict with the deployed databases? I'm doing this locally, dropping, but then migrations will be run on deployed application, which still has old migrations table. Will this be a problem?

**Reply by B4nan** [reply]
You'd have to adjust the migrations table in production manually I guess. 

Personally I would keep the very last executed migration and alter its contents, that way you wont need to deal with this. You don't really need to wipe things and use the `--initial` flag, as I mentioned above, it will use the output of `schema:create --dump`, so you can just do that to get the schema snapshot, put it into the last executed migration and remove all the old ones. It shouldn't care about old migration files being missing.
